sklearn/preprocessing/label.py

```diff
@@ -33,131 +33,217 @@
+                if len(unseen):
+                    raise ValueError("y contains previously unseen labels: %s"
+                                     % str(unseen), unseen)
 
-    Returns
-    -------
-    diff : list
-        The unique values present in `values` and not in `uniques` (the
-        unknown values).
-    valid_mask : boolean array
-        Additionally returned if ``return_mask=True``.
+        encoded = np.searchsorted(uniques, values)
+        from ..impute import _get_mask
+        missing_mask = _get_mask(values, missing_values)
 
-    """
-    if values.dtype == object:
-        uniques_set = set(uniques)
-        diff = list(set(values) - uniques_set)
-        if return_mask:
-            if diff:
-                valid_mask = np.array([val in uniques_set for val in values])
+        if encode_unknown:
+            if check_values:
+                unknown_mask = _nanin1d(values, unseen)
             else:
-                valid_mask = np.ones(len(values), dtype=bool)
-            return diff, valid_mask
+                unknown_mask = np.zeros_like(values, dtype=np.bool)
+            return uniques, encoded, missing_mask, unknown_mask
         else:
-            return diff
+            return uniques, encoded, missing_mask
     else:
-        unique_values = np.unique(values)
-        diff = list(np.setdiff1d(unique_values, uniques, assume_unique=True))
-        if return_mask:
-            if diff:
-                valid_mask = np.in1d(values, uniques)
-            else:
-                valid_mask = np.ones(len(values), dtype=bool)
-            return diff, valid_mask
+        return uniques
+
+
+# Since it is more efficient to remove an item from a set, object exclusion
+# in included in the _nanunique_object. Also since None and np.nan are not
+# sortable, they are handled separately after the sort.
+def _nanunique_object(ar, exclude_value):
+    items = set(ar)
+
+    # nan might not be discarded since nan comes in different forms
+    items.discard(exclude_value)
+
+    # Handle None afterwards if it is in the items. Set a flag for now.
+    try:
+        items.remove(None)
+        has_none = True
+    except KeyError:
+        has_none = False
+
+    # Handle nan afterwards if it is in the items. Set a flag for now. Since
+    # nan can come in different forms, we check everything.
+    items_not_na = [i for i in items if not is_scalar_nan(i)]
+    has_na = len(items) > len(items_not_na)
+
+    # Sort without None and nan
+    uniques = sorted(items_not_na)
+
+    # Bring back None if needed
+    if has_none:
+        uniques.append(None)
+
+    # Being back nan if needed. Since nan comes in different forms, nan might
+    # exists despite discarding it from the set
+    if has_na and not is_scalar_nan(exclude_value):
+        uniques.append(np.nan)
+
+    return uniques
+
+
+# Since nan comes in multiple forms, hash is not enough to identify it
+def _dict_to_mapper(d, **kwargs):
```

Comment:
FWIW, unless I'm much mistaken, this is similar to the functionality provided by:

```
class nanmapper(dict):
    def __missing__(self, key):
        return nan_value
```